ocaml/idl/latex_backend.ml

```diff
@@ -123,253 +83,206 @@ let flatten stuff =
-      "\\hline";
-      "Quals & Field & Type & Description \\\\";
-      "\\hline"
-    ]
-
-
-let class_footer =
-  [
-    "\\hline";
-    "\\end{longtable}"
-  ]
-
-
-let field_table_of_obj newpage x closed =
-  let field_tex = List.concat (List.map (fun x -> of_content (flatten x) closed) x.contents) in
-  (if newpage then ["\\newpage"] else []) @
-  [
-    "\\subsection{Fields for class: "^(escape x.name)^"}";
-  ] @
-  (if x.contents=[] then
-     ["{\\bf Class "^(escape x.name)^" has no fields.}"]
-   else
-     (class_header x closed) @ field_tex @ class_footer)
-
-let of_obj x closed =
-  [
-    "\\newpage";
-    "\\section{Class: "^(escape x.name)^"}" ]
-  @ (field_table_of_obj false x closed)
-  @
-  [
-    "\\subsection{RPCs associated with class: "^(escape x.name)^"}"
-  ]
-  @
-  (if x.messages=[] then
-     ["\n\n";
-      "{\\bf Class "^(escape x.name)^" has no additional RPCs associated with it.}"]
-   else
-     [
-       latex_of_funblock closed "sub" x.messages
-     ])
-
-let error_signature name params =
+    List.iter (latex_section_of_message printer) x.messages
+
+let print_enum printer = function
+  | Enum (name, options) ->
+    printer (sprintf "|`enum %s`||" name);
+    printer "|---|---|";
+    let print_option (opt, description) = printer (sprintf "|`%s`|%s|" opt (escape description)) in
+    List.iter print_option options;
+    printer "";
+  | _ -> ()
+
+let error_doc printer { err_name=name; err_params=params; err_doc=doc } =
+  printer (sprintf "#### %s" (escape name));
+  printer "";
+  printer (escape doc);
+  printer "";
   if params = [] then
-    "
-\\vspace{0.3cm}
-No parameters."
-  else
-    sprintf "
-\\vspace{0.3cm}
-{\\bf Signature:}
-\\begin{verbatim}%s(%s)\\end{verbatim}"
-      name (String.concat ", " params)
-
-let error_doc { err_name=name; err_params=params; err_doc=doc } =
-  printf "
-\\subsubsection{%s}
-
-%s%s
-\\begin{center}\\rule{10em}{0.1pt}\\end{center}
-" (escape name) (wrap (escape doc)) (error_signature name params)
-
-let include_file ?(escaped=false) ?(blanklines=false) filename =
-  let ic = open_in filename in
-  try
-    while true do
-      let line = input_line ic in
-      print_endline (if escaped then escape line else line);
-      if blanklines then print_endline "";
-    done
-  with End_of_file -> ()
-
-and error_section () =
-  print_endline "\\newpage
-\\section{Error Handling}
+    printer "No parameters."
+  else begin
+    printer "_Signature:_";
+    printer "```";
+    printer (sprintf "%s(%s)" name (String.concat ", " params));
+    printer "```"
+  end;
+  printer ""
+
+let print_all printer api =
+  (* Remove private messages that are only used internally (e.g. get_record_internal) *)
+  let api = Dm_api.filter (fun _ -> true) (fun _ -> true)
+      (fun msg -> match msg.msg_tag with (FromObject (Private _)) -> false | _ -> true) api in
```

Comment:
Just one optional comment: The generated API reference includes the [VMPP](https://xapi-project.github.io/xen-api/classes/vmpp.html) class, but that class has been removed. However, this is a bigger issue, that also affects https://xapi-project.github.io/xen-api/ (the class shows up in the menu, but I think it shouldn't, as it just takes up space), and we still generate interfaces for it, and as a result, we have to keep https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_vmpp.ml around. So probably filtering out the removed elements is out of scope for this PR, or we could filter out removed things here in `print_all` the same way we filter out private things.